modules/dcache/src/main/java/org/dcache/pool/statistics/IoStatistics.java

```diff
@@ -42,20 +46,25 @@
      *         The duration of the read request in nanoseconds; must be non-negative
      */
 
-    public void updateRead(long readBytes, long readTime) {
+    public void updateRead(long readBytes, long readTime, long requestedBytes) {
 
-        double readSpeed = calculateSpeed(readBytes, readTime);
-        _readSpeeds[_readSpeedsArrayIndex] = readSpeed;
-        _readSpeedsArrayIndex = (_readSpeedsArrayIndex + 1) % _readSpeeds.length;
+        if(readBytes >= 0) {
+            double readSpeed = calculateSpeed(readBytes, readTime);
+            _readSpeeds[_readSpeedsArrayIndex] = readSpeed;
+            _readSpeedsArrayIndex = (_readSpeedsArrayIndex + 1) % _readSpeeds.length;
 
-        _minReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.min(readSpeed, _minReadSpeed);
-        _maxReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.max(readSpeed, _maxReadSpeed);
-        _avgReadSpeed = (_readRequestNum == 0) ? readSpeed : calculateNewAvg(_readRequestNum,_avgReadSpeed, readSpeed);
-        _95ReadSpeed = (_readRequestNum == 0) ? readSpeed : percentile.evaluate(_readSpeeds, 0, (_readSpeedsArrayIndex - 1));
+            _minReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.min(readSpeed, _minReadSpeed);
```

Comment:
Actually, I misspoke.  the percentile.evaluate should work on the first call because (a) you already added a value to the array and (b) you already incremented your index.  The second value should be the effective length of the array, not the last index, BTW.

percentile.evaluate(_readSpeeds, 0, _readSpeedsArrayIndex) is correct, I think.  In the case of the first call, you would have an array of length 1, and that is the third parameter:

--> evaluate(double[] values, int start, int length)

The problem comes when your index reaches length and turns over.  Your computation then will get messed up because you are calling the modulo index length.  

So we need to do something else there.  Perhaps using the _request counter and doing

percentile.evaluate(_readSpeeds, 0, Math.min(_readSpeeds.length, _readRequestNum) ?